### PR TITLE
Fix: Handle new lines inside block string correctly

### LINKED ISSUE #1636: [lexer] Block string does not advance line
Lexer does not advance line number when tokenizing block strings.

Line numbers are then shifted on all following tokens

### PR DESCRIPTION
**Pull request by langpavel**
This fixes #1636

### CODE REVIEW

**Comment by langpavel** [comment]
Hi @IvanGoncharov 
Can I kindly ask you for review?
Thanks!

File: src/language/lexer.js [hunk 1]
```diff
@@ -668,8 +668,22 @@ function readBlockString(source, start, line, col, prev): Token {
...
 
-    // Escape Triple-Quote (\""")
-    if (
+    if (code === 10) {
+      // new line
+      ++position;
+      ++lexer.line;
+      lexer.lineStart = position;
+    } else if (code === 13) {
+      // carriage return
```
Comment (IvanGoncharov) [inline src/language/lexer.js]: @langpavel Can you please add test cases that cover `\r\n` and `\n\r`?

File: src/language/__tests__/lexer-test.js [hunk 2]
```diff
@@ -359,6 +365,24 @@ describe('Lexer', () => {
...
 
+  it('advance line after lexing block string', () => {
+    expect(
+      lexSecond(`"""
+
+        spans
+          multiple
+            lines
+
+        """ second_token`),
```
Comment (IvanGoncharov) [inline src/language/__tests__/lexer-test.js]: To cover all cases in one test you can do something like this:
```js
const str = [
  '""" \n',
  'spans \r\n',
  'multiple \n\r',
  'lines \n\n',
  '""" second_token'
].join('')
```

**Comment by IvanGoncharov** [comment]
@langpavel Thanks for PR 👍 
Please expand tests and I will merge this PR.

P.S. It's a good temporary solution but I don't like duplication of new line code handling and having special handling of block strings.
I think in long run we need to implement a common function for advancing lexer to the next character. Maybe even unify with `CharacterStream` from language server:
https://github.com/graphql/graphql-language-service/blob/master/packages/parser/src/CharacterStream.js

**Comment by langpavel** [comment]
Ok, I will try refactor to new lexer method `advanceLine`, existing tests should cover this so no need to expand tests for block string

**Comment by langpavel** [comment]
@IvanGoncharov I extracted `consumeNewline` function so newlines are handled on one place.

File: src/language/lexer.js [hunk 3]
```diff
@@ -361,6 +361,34 @@ function unexpectedCharacterMessage(code) {
   return `Cannot parse the unexpected character ${printCharCode(code)}.`;
 }
 
+function consumeNewline(
+  startPosition: number,
+  code: number,
+  codeahead: number,
+  lexer: Lexer<*>,
+): [boolean, number] {
```
Comment (IvanGoncharov) [inline src/language/lexer.js]: @langpavel It complicates code :(
Just revert to previous commit.
My point was that sometime in future we need to refactor Lexer, and your previous code was good until then.

Comment (langpavel) [reply src/language/lexer.js]: Ok, I will force push up to previous commit

Comment (IvanGoncharov) [reply src/language/lexer.js]: @langpavel Sorry for confusing you

Comment (langpavel) [reply src/language/lexer.js]: Last commit deleted

**Comment by IvanGoncharov** [comment]
> existing tests should cover this so no need to expand tests for block string

@langpavel Existing tests don't check that `\r\n` or `\n\r` advance line counter by one.

File: src/language/__tests__/lexer-test.js (hunk 2 above)
Comment (langpavel) [reply src/language/__tests__/lexer-test.js]: This case will not cover edge cases because various combinations can annihilate

**Comment by langpavel** [comment]
Ok, I will add more tests

**Comment by langpavel** [comment]
@IvanGoncharov I added your test case and also added `column` assertion

File: src/language/lexer.js (hunk 1 above)
Comment (langpavel) [reply src/language/lexer.js]: Added

File: src/language/__tests__/lexer-test.js (hunk 2 above)
Comment (langpavel) [reply src/language/__tests__/lexer-test.js]: Added, exactly like you suggested

**Comment by IvanGoncharov** [comment]
@langpavel Thanks 👍 
Merged 🎉

**Comment by langpavel** [comment]
Thanks! Happy new year :tada:
